Added usersync support to Between SSP adapter; Major fixes and refactor

egsk opened the pull request:
Hello! Unfortunately, Between Adapter is not working currently. Previous contributor forgot about some SSP-specific parameters. I rewrote adapter, added usersync support, added missing parameters and [tested it with real-world example](https://egorskorokhodov.ru/pbs.html).
Test coverage more than 90%.

SyntaxNode commented:
Thank you for the refactor. We're a bit busy at the moment so it may take us longer than usual to review. Please stay patient. :)

mansinahar commented on adapters/between/between.go:
```diff
@@ -3,169 +3,223 @@ package between
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "ignoring BidRequest: no valid Imps",
```
Please consider changing this message to saying `Invalid Bidder ext` or something along those lines. Similar to what you have below on L83

egsk replied:
fixed

mansinahar commented on adapters/between/between.go:
```diff
@@ -3,169 +3,223 @@ package between
+func addImpProps(imp *openrtb.Imp, secure *int8, betweenExt *openrtb_ext.ExtImpBetween) {
+	imp.Secure = secure
+	bidFloor, err := strconv.ParseFloat(betweenExt.BidFloor, 64)
+	if err != nil {
```
Please consider simplifying this as follows:
```
if bidFloor, err := strconv.ParseFloat(betweenExt.BidFloor, 64); err != nil || bidFloor <= 0 {
	imp.BidFloor = defaultBidfloor
} else {
	imp.BidFloor = bidFloor
}
```

egsk replied:
fixed

mansinahar commented on adapters/between/between.go:
```diff
@@ -3,169 +3,223 @@ package between
 		if err != nil {
 			errors = append(errors, err)
-			continue
+			return nil, errors
```
Are you sure you want to return at this point even if there was an error in unpacking only one of the imp ext? The other imps could still be valid and in that case you'd be throwing away those too. It might make more sense to append the error as you're doing now and ignore this imp but continue with the rest of them

egsk replied:
fixed

mansinahar commented on adapters/between/between.go:
```diff
@@ -3,169 +3,223 @@ package between
+
+		if err := buildImpBanner(&imp); err != nil {
+			errors = append(errors, err)
+			return nil, errors
```
Similar comment as above

egsk commented on adapters/between/between.go:
```diff
@@ -3,169 +3,223 @@ package between
-	// Pull the host info from the bidder params.
-	reqImps, errs := splitImpressions(request.Imp)
+const (
+	defaultBidfloor = 0.00001 // BetweenSSP requires bidfloor > 0. If BidFloor of openrtb_ext.ExtImpBetween is zero, set it to defaultBidFloor value, see addImpProps
```
fixed :)

egsk commented:
Thanks for the feedback! Everything has been fixed

mansinahar commented on usersync/usersyncers/syncer.go:
```diff
@@ -164,6 +165,7 @@ func NewSyncerMap(cfg *config.Configuration) map[openrtb_ext.BidderName]usersync
 	insertIntoMap(cfg, syncers, openrtb_ext.BidderYieldmo, yieldmo.NewYieldmoSyncer)
 	insertIntoMap(cfg, syncers, openrtb_ext.BidderYieldone, yieldone.NewYieldoneSyncer)
 	insertIntoMap(cfg, syncers, openrtb_ext.BidderZeroClickFraud, zeroclickfraud.NewZeroClickFraudSyncer)
+	insertIntoMap(cfg, syncers, openrtb_ext.BidderBetween, between.NewBetweenSyncer)
```
Now that you are adding the usersync capability, please remove `Between` from the `adaptersWithoutSyncers` map in the `syncer_test.go` file

mansinahar commented on openrtb_ext/imp_between.go:
```diff
@@ -1,5 +1,8 @@
-	Host string `json:"host"`
+	Host        string `json:"host"`
+	PublisherID string `json:"publisher_id"`
+	BidFloor    string `json:"bid_floor"`
```
Are `bid_floor` and `bid_floor_cur` also params that you expect in the imp ext? Are they just optional parameters that you pass it to your bidder as is if present? If so, please add them to the `static/bidder-params/between.json` file as well as optional parameters

mansinahar commented on openrtb_ext/imp_between.go:
```diff
@@ -1,5 +1,8 @@
 type ExtImpBetween struct {
-	Host string `json:"host"`
+	Host        string `json:"host"`
+	PublisherID string `json:"publisher_id"`
```
Please add `publisher_id` to `static/bidder-params/between.json` as a required parameter

mansinahar commented on adapters/between/between.go:
```diff
@@ -3,169 +3,219 @@ package between
+	var betweenExt *openrtb_ext.ExtImpBetween
+	for _, imp := range request.Imp {
+		var err error
+		betweenExt, err = unpackImpExt(&imp)
```
This would mean that you only consider the `ext` of the last imp in the request and use the values set in that to build the endpoint URL. So if the first imp has say an `ext` with `host` set as `127.0.0.1` and the next one has `host` set as `127.0.0.2` then you'd consider `127.0.0.2` as the host when building the URL. Are you sure you're okay with this behavior?

mansinahar commented:
> Thanks for the feedback! Everything has been fixed

Thanks for the fixes @egsk. There are a few more comments. if you could please address those too that'd be great. There were some things that I missed in my earlier review so sorry about that.

guscarreon commented:
@egsk your adapter looks pretty good, but starting our next release on December 3rd, adapter source code will need to be written in a slightly different way. Pull request #1532 describes this changes. Therefore, before we merge your adapter please take a look to PR #1532 and modify your adapter accordingly. If you have any questions, we're glad to help.

SyntaxNode commented:
We also just published a [new adapter doc](https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html) today which will guide you through the new approach.

mansinahar commented on adapters/between/between.go:
```diff
@@ -3,162 +3,214 @@ package between
-			continue
+	var betweenExt openrtb_ext.ExtImpBetween
+	if err := json.Unmarshal(bidderExt.Bidder, &betweenExt); err != nil {
+		fmt.Println(err)
```
Looks like a leftover `fmt.Println`. Can you please remove this line?

mansinahar reviewed (commented):
Other than this minor comment, I think this LGTM. One last thing I noticed is that the `Between` adapter doesn't have a `params_test.go` file that tests parameter validation and I'd request you to consider adding one. 

For reference, see this `params_test.go` file for `Adform`: https://github.com/prebid/prebid-server/blob/master/adapters/adform/params_test.go

egsk commented:
> Other than this minor comment, I think this LGTM. One last thing I noticed is that the `Between` adapter doesn't have a `params_test.go` file that tests parameter validation and I'd request you to consider adding one.
> 
> For reference, see this `params_test.go` file for `Adform`: https://github.com/prebid/prebid-server/blob/master/adapters/adform/params_test.go

Thansk for the feedback! I have added params test and fixed other comments

mansinahar reviewed (approved):
LGTM!! Thanks for addressing all the feedback :)

mansinahar commented:
@egsk Can you also please create a docs PR reflecting the parameter changes in the [prebid/prebid.github.io](https://github.com/prebid/prebid.github.io) repo: https://github.com/prebid/prebid.github.io/blob/master/dev-docs/bidders/between.md

egsk commented:
> @egsk Can you also please create a docs PR reflecting the parameter changes in the [prebid/prebid.github.io](https://github.com/prebid/prebid.github.io) repo: https://github.com/prebid/prebid.github.io/blob/master/dev-docs/bidders/between.md ?

Thanks for your attention :)
PR reflecting the parameter changes: https://github.com/prebid/prebid.github.io/pull/2549

guscarreon reviewed (approved):
LGTM
